activerecord/lib/active_record/attributes.rb

```diff
@@ -205,14 +205,21 @@ module ClassMethods
       # tracking is performed. The methods +changed?+ and +changed_in_place?+
       # will be called from ActiveModel::Dirty. See the documentation for those
       # methods in ActiveModel::Type::Value for more details.
-      def attribute(name, cast_type = nil, **options, &block)
+      def attribute(name, cast_type = nil, **options, &_decorator)
         name = name.to_s
         reload_schema_from_cache
 
-        self.attributes_to_define_after_schema_loads =
-          attributes_to_define_after_schema_loads.merge(
-            name => [cast_type || block, options]
-          )
+        prev_cast_type, prev_options, prev_decorator = attributes_to_define_after_schema_loads[name]
+
+        unless cast_type && prev_cast_type
+          cast_type ||= prev_cast_type
+          options = prev_options || options if options.empty?
+          _decorator ||= prev_decorator
+        end
+
+        self.attributes_to_define_after_schema_loads = attributes_to_define_after_schema_loads.merge(
+          name => [cast_type, options, _decorator]
```

Comment:
Don't think we need the `_` prefix on the decorator, we won't clash with the instance method here.

Reply:
The underscore was to mark the block as a private parameter (based on https://github.com/rails/rails/pull/39929#discussion_r461361532), but I think 5a592cf620a3770f697388e347137d85d7592ac9 takes care of that now. :+1: